File: 76.minWindow.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> hash;
        int num = t.size();
        int len = INT_MAX;
        int start = 0;
        int left = 0;

        for(auto val: t) hash[val]++;

        for(int i =0; i < s.size(); i++) {
            if(hash[s[i]]-- >0) num--;
            while(num==0) {
                if (len>(i-left+1)) {
                    len = i - left + 1;
                    start = left;
                }
                if (hash[s[left]] == 0) {
                    num++;
                }
                hash[s[left]]++;
                left++;
            }
        }

        if (len == INT_MAX) return "";
        return s.substr(start, len);
    }
};

File: 76.minWindow.go
func minWindow(s string, t string) string {
    cnt := make(map[byte]int)
    remain := len(t)
    length := int(^uint(0) >> 1)
    left := 0

    start := -1
    end := 0
	
	for i := 0; i < len(t); i++ {
		cnt[t[i]]++;
	}


    for i := 0; i < len(s); i++ {
        cnt[s[i]]--;
        if cnt[s[i]] >= 0 {
            remain--;
        }
        
        for remain == 0 {
            if length > i - left + 1 {
                length = i - left + 1
                start = left
                end = i
            }

            if cnt[s[left]] == 0 {
                remain++
            }
            cnt[s[left]]++
            left++
        }
    }

    if start == -1 {return ""}
    return s[start:end+1]
}